mail_embed_image/models/ir_mail_server.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2019 Therp BV <https://therp.nl>
+# Copyright 2024 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+import uuid
+from base64 import b64encode
+from email.mime.image import MIMEImage
+
+import requests
+from lxml.html import fromstring, tostring
+
+from odoo import models
+
+_logger = logging.getLogger(__name__)
+
+
+class IrMailServer(models.Model):
+    _inherit = "ir.mail_server"
+
+    def build_email(
+        self,
+        email_from,
+        email_to,
+        subject,
+        body,
+        email_cc=None,
+        email_bcc=None,
+        reply_to=False,
+        attachments=None,
+        message_id=None,
+        references=None,
+        object_id=False,
+        subtype="plain",
+        headers=None,
+        body_alternative=None,
+        subtype_alternative="plain",
+    ):
+        image_embedding_method = self.env.company.image_embedding_method
+        fileparts = None
+        if subtype == "html" and image_embedding_method != "none":
+            body, fileparts = self._build_email_replace_img_src(body)
+        result = super(IrMailServer, self).build_email(
+            email_from=email_from,
+            email_to=email_to,
+            subject=subject,
+            body=body,
+            email_cc=email_cc,
+            email_bcc=email_bcc,
+            reply_to=reply_to,
+            attachments=attachments,
+            message_id=message_id,
+            references=references,
+            object_id=object_id,
+            subtype=subtype,
+            headers=headers,
+            body_alternative=body_alternative,
+            subtype_alternative=subtype_alternative,
+        )
+        if fileparts:
+            for fpart in fileparts:
+                result.attach(fpart)
+            # Multipart method MUST be multipart/related for CIDs embedding
+            # Gmail and Office won't process the images otherwise
+            if image_embedding_method == "cid":
+                result.set_type("multipart/related")
+        return result
+
+    def _build_email_replace_img_src(self, html_body):
+        """Replace img src with base64 encoded image."""
+        if not html_body:
+            return html_body
+
+        base_url = self.env["ir.config_parameter"].get_param("web.base.url")
+        image_embedding_method = self.env.company.image_embedding_method
+        root = fromstring(html_body)
+        fileparts = []
+        # Limit results to only internal resources to avoid malicious external
+        # image injections
+        for img in root.xpath(
+            ".//img[starts-with(@src, '%s')]"
+            "| .//img[starts-with(@src, '/web/image')]" % (base_url)
+        ):
+            image_path = img.get("src")
+            try:
+                response = requests.get(image_path, timeout=10)
```

Comment:
please add a test case where we have an image with a relative url, I believe this will fail here